homeassistant/components/esphome/light.py

```diff
@@ -59,20 +54,76 @@ async def async_setup_entry(
     )
 
 
-_COLOR_MODES: EsphomeEnumMapper[LightColorMode, str] = EsphomeEnumMapper(
-    {
-        LightColorMode.UNKNOWN: COLOR_MODE_UNKNOWN,
-        LightColorMode.ON_OFF: COLOR_MODE_ONOFF,
-        LightColorMode.BRIGHTNESS: COLOR_MODE_BRIGHTNESS,
-        LightColorMode.WHITE: COLOR_MODE_WHITE,
-        LightColorMode.COLOR_TEMPERATURE: COLOR_MODE_COLOR_TEMP,
-        LightColorMode.COLD_WARM_WHITE: COLOR_MODE_COLOR_TEMP,
-        LightColorMode.RGB: COLOR_MODE_RGB,
-        LightColorMode.RGB_WHITE: COLOR_MODE_RGBW,
-        LightColorMode.RGB_COLOR_TEMPERATURE: COLOR_MODE_RGBWW,
-        LightColorMode.RGB_COLD_WARM_WHITE: COLOR_MODE_RGBWW,
-    }
-)
+_COLOR_MODE_MAPPING = {
+    COLOR_MODE_ONOFF: [
+        LightColorCapability.ON_OFF,
+    ],
+    COLOR_MODE_BRIGHTNESS: [
+        LightColorCapability.ON_OFF | LightColorCapability.BRIGHTNESS,
+        # for compatibility with older clients
+        LightColorCapability.BRIGHTNESS,
+    ],
+    COLOR_MODE_COLOR_TEMP: [
+        LightColorCapability.ON_OFF
+        | LightColorCapability.BRIGHTNESS
+        | LightColorCapability.COLOR_TEMPERATURE,
+        LightColorCapability.ON_OFF
+        | LightColorCapability.BRIGHTNESS
+        | LightColorCapability.COLD_WARM_WHITE,
+    ],
+    COLOR_MODE_RGB: [
+        LightColorCapability.ON_OFF
+        | LightColorCapability.BRIGHTNESS
+        | LightColorCapability.RGB,
+    ],
+    COLOR_MODE_RGBW: [
+        LightColorCapability.ON_OFF
+        | LightColorCapability.BRIGHTNESS
+        | LightColorCapability.RGB
+        | LightColorCapability.WHITE,
+    ],
+    COLOR_MODE_RGBWW: [
+        LightColorCapability.ON_OFF
+        | LightColorCapability.BRIGHTNESS
+        | LightColorCapability.RGB
+        | LightColorCapability.COLOR_TEMPERATURE,
+        LightColorCapability.ON_OFF
+        | LightColorCapability.BRIGHTNESS
+        | LightColorCapability.RGB
+        | LightColorCapability.COLD_WARM_WHITE,
+    ],
+    COLOR_MODE_WHITE: [LightColorCapability.ON_OFF | LightColorCapability.WHITE],
+}
+
+
+def _color_mode_to_ha(mode: int) -> str:
+    """Convert an esphome color mode to a HA color mode constant.
+
+    Choses the color mode that best matches the feature-set.
+    """
+    candidates = []
+    for ha_mode, cap_lists in _COLOR_MODE_MAPPING.items():
+        for caps in cap_lists:
+            if caps == mode:
+                # exact match
+                return ha_mode
+            if (mode & caps) == caps:
+                # all requirements met
+                candidates.append((ha_mode, caps))
+
+    if not candidates:
+        return COLOR_MODE_UNKNOWN
```

Comment:
Is this needed, can an ESPHome light be setup such that it doesn't even support on/off and if it can do we need to support that?

Reply:
Right now, not in the core integrations. But in theory a custom integration could advertise such a color mode, this ensures we return _something_ in that case.